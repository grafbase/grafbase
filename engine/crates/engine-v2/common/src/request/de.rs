use serde::{
    de::{DeserializeSeed, Visitor},
    Deserialize, Deserializer,
};
use std::borrow::Cow;

use crate::{BorrowedValue, BorrowedVariables};

pub(super) fn deserialize_non_empty_vec<'de, D, T>(deserializer: D) -> Result<Vec<T>, D::Error>
where
    D: Deserializer<'de>,
    T: Deserialize<'de>,
{
    use serde::de::Error as _;

    let v = <Vec<T>>::deserialize(deserializer)?;
    if v.is_empty() {
        Err(D::Error::invalid_length(0, &"a non-empty sequence"))
    } else {
        Ok(v)
    }
}

impl<'de: 'a, 'a> Deserialize<'de> for BorrowedVariables<'a> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let mut seed = BorrowedVariablesSeed::default();
        let root = deserializer.deserialize_any(&mut seed)?;
        if !matches!(root, BorrowedValue::Null | BorrowedValue::Map(_)) {
            return Err(serde::de::Error::custom("variables must be a map or null"));
        }
        Ok(BorrowedVariables {
            root,
            values: seed.values,
            key_values: seed.key_values,
        })
    }
}

#[derive(Default)]
struct BorrowedVariablesSeed<'a> {
    values: Vec<BorrowedValue<'a>>,
    values_buffer_pool: Vec<Vec<BorrowedValue<'a>>>,
    key_values: Vec<(Cow<'a, str>, BorrowedValue<'a>)>,
    key_values_buffer_pool: Vec<Vec<(Cow<'a, str>, BorrowedValue<'a>)>>,
}

impl<'de> DeserializeSeed<'de> for &mut BorrowedVariablesSeed<'de> {
    type Value = BorrowedValue<'de>;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_option(self)
    }
}

impl<'de> Visitor<'de> for &mut BorrowedVariablesSeed<'de> {
    type Value = BorrowedValue<'de>;

    fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str("a JSON value")
    }

    fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_any(self)
    }

    fn visit_none<E>(self) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::Null)
    }

    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
    where
        A: serde::de::MapAccess<'de>,
    {
        let mut buffer = self.key_values_buffer_pool.pop().unwrap_or_default();

        while let Some(key) = map.next_key::<Cow<'de, str>>()? {
            let value: BorrowedValue<'de> = map.next_value_seed(&mut *self)?;
            buffer.push((key, value));
        }

        let start = self.key_values.len();
        buffer.sort_unstable_by(|a, b| a.0.cmp(&b.0));
        self.key_values.append(&mut buffer);
        self.key_values_buffer_pool.push(buffer);
        let end = self.key_values.len();

        Ok(BorrowedValue::Map((start..end).into()))
    }

    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
    where
        A: serde::de::SeqAccess<'de>,
    {
        let mut buffer = self.values_buffer_pool.pop().unwrap_or_default();

        while let Some(value) = seq.next_element_seed(&mut *self)? {
            buffer.push(value);
        }

        let start = self.values.len();
        self.values.append(&mut buffer);
        self.values_buffer_pool.push(buffer);
        let end = self.values.len();

        Ok(BorrowedValue::List((start..end).into()))
    }

    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::String(v.to_string().into()))
    }

    fn visit_borrowed_str<E>(self, v: &'de str) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::String(v.into()))
    }

    fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::String(v.into()))
    }

    fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::Bool(v))
    }

    fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::I64(v))
    }

    fn visit_i32<E>(self, v: i32) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::I64(v as i64))
    }

    fn visit_i16<E>(self, v: i16) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::I64(v as i64))
    }

    fn visit_i8<E>(self, v: i8) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::I64(v as i64))
    }

    fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::U64(v))
    }

    fn visit_u32<E>(self, v: u32) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::U64(v as u64))
    }

    fn visit_u16<E>(self, v: u16) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::U64(v as u64))
    }

    fn visit_u8<E>(self, v: u8) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::U64(v as u64))
    }

    fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::F64(v))
    }

    fn visit_f32<E>(self, v: f32) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        Ok(BorrowedValue::F64(v as f64))
    }
}
