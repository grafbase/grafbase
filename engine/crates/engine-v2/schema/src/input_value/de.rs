use serde::{
    de::{
        value::{MapDeserializer, SeqDeserializer},
        IntoDeserializer, Visitor,
    },
    forward_to_deserialize_any,
};

use crate::{InputValue, InputValueSerdeError, InputValuesContext};

pub(super) struct InputValueDeserializer<'de, Str, Ctx> {
    pub ctx: Ctx,
    pub value: &'de InputValue<Str>,
}

impl<'de, Str, Ctx> serde::Deserializer<'de> for InputValueDeserializer<'de, Str, Ctx>
where
    Ctx: InputValuesContext<'de, Str>,
    Str: 'de,
{
    type Error = InputValueSerdeError;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        match self.value {
            InputValue::Null => visitor.visit_unit(),
            InputValue::String(s) | InputValue::UnknownEnumValue(s) => visitor.visit_borrowed_str(self.ctx.get_str(s)),
            InputValue::EnumValue(id) => visitor.visit_borrowed_str(self.ctx.schema_walker().walk(*id).name()),
            InputValue::Int(n) => visitor.visit_i32(*n),
            InputValue::BigInt(n) => visitor.visit_i64(*n),
            InputValue::U64(n) => visitor.visit_u64(*n),
            InputValue::Float(n) => visitor.visit_f64(*n),
            InputValue::Boolean(b) => visitor.visit_bool(*b),
            &InputValue::List(ids) => {
                let ctx = self.ctx;
                let mut deserializer = SeqDeserializer::new(ids.iter().map(move |id| InputValueDeserializer {
                    value: &ctx.input_values()[id],
                    ctx,
                }));
                let seq = visitor.visit_seq(&mut deserializer)?;
                deserializer.end()?;
                Ok(seq)
            }
            InputValue::InputObject(ids) => {
                let ctx = self.ctx;
                let mut deserializer = MapDeserializer::new(ids.iter().map(move |id| {
                    let (id, value) = &ctx.input_values()[id];
                    (
                        ctx.schema_walker().walk(*id).name(),
                        InputValueDeserializer { value, ctx },
                    )
                }));
                let map = visitor.visit_map(&mut deserializer)?;
                deserializer.end()?;
                Ok(map)
            }
            InputValue::Map(ids) => {
                let ctx = self.ctx;
                let mut deserializer = MapDeserializer::new(ids.iter().map(move |id| {
                    let (key, value) = &ctx.input_values()[id];
                    (self.ctx.get_str(key), InputValueDeserializer { value, ctx })
                }));
                let map = visitor.visit_map(&mut deserializer)?;
                deserializer.end()?;
                Ok(map)
            }
        }
    }

    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        if matches!(self.value, InputValue::Null) {
            visitor.visit_none()
        } else {
            visitor.visit_some(self)
        }
    }

    forward_to_deserialize_any! {
        bool i8 i16 i32 i64 i128 u8 u16 u32 u64 u128 f32 f64 char str string
        bytes byte_buf unit unit_struct newtype_struct seq tuple
        tuple_struct map struct enum identifier ignored_any
    }
}

impl<'de, Str, Ctx> IntoDeserializer<'de, InputValueSerdeError> for InputValueDeserializer<'de, Str, Ctx>
where
    Ctx: InputValuesContext<'de, Str>,
    Str: 'de,
{
    type Deserializer = Self;
    fn into_deserializer(self) -> Self::Deserializer {
        self
    }
}
