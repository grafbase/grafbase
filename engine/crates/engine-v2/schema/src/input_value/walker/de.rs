use serde::{
    de::{
        value::{MapDeserializer, SeqDeserializer},
        IntoDeserializer, Visitor,
    },
    forward_to_deserialize_any,
};
use walker::Walk;

use crate::{InputValueSerdeError, SchemaInputValueRecord};

use super::SchemaInputValue;

impl<'de> serde::Deserializer<'de> for SchemaInputValue<'de> {
    type Error = InputValueSerdeError;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        let SchemaInputValue { schema, value } = self;
        match value {
            SchemaInputValueRecord::Null => visitor.visit_none(),
            SchemaInputValueRecord::String(id) | SchemaInputValueRecord::UnboundEnumValue(id) => {
                visitor.visit_borrowed_str(&schema[*id])
            }
            SchemaInputValueRecord::EnumValue(id) => visitor.visit_borrowed_str(id.walk(schema).name()),
            SchemaInputValueRecord::Int(n) => visitor.visit_i32(*n),
            SchemaInputValueRecord::BigInt(n) => visitor.visit_i64(*n),
            SchemaInputValueRecord::U64(n) => visitor.visit_u64(*n),
            SchemaInputValueRecord::Float(n) => visitor.visit_f64(*n),
            SchemaInputValueRecord::Boolean(b) => visitor.visit_bool(*b),
            SchemaInputValueRecord::List(ids) => SeqDeserializer::new(ids.walk(schema)).deserialize_any(visitor),
            SchemaInputValueRecord::InputObject(ids) => MapDeserializer::new(
                ids.walk(schema)
                    .map(|(input_value_definition, value)| (input_value_definition.name(), value)),
            )
            .deserialize_any(visitor),
            SchemaInputValueRecord::Map(ids) => MapDeserializer::new(ids.walk(schema)).deserialize_any(visitor),
        }
    }

    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        if matches!(self.value, SchemaInputValueRecord::Null) {
            visitor.visit_none()
        } else {
            visitor.visit_some(self)
        }
    }

    fn deserialize_ignored_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        visitor.visit_unit()
    }

    forward_to_deserialize_any! {
        bool i8 i16 i32 i64 i128 u8 u16 u32 u64 u128 f32 f64 char str string
        bytes byte_buf unit unit_struct newtype_struct seq tuple
        tuple_struct map struct enum identifier
    }
}

impl<'de> IntoDeserializer<'de, InputValueSerdeError> for SchemaInputValue<'de> {
    type Deserializer = Self;
    fn into_deserializer(self) -> Self::Deserializer {
        self
    }
}
