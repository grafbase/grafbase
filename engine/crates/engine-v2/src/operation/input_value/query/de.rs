use schema::InputValueSerdeError;
use serde::{
    de::{
        value::{MapDeserializer, SeqDeserializer},
        IntoDeserializer, Visitor,
    },
    forward_to_deserialize_any,
};

use super::{QueryInputValue, QueryInputValueWalker};

impl<'de> serde::Deserializer<'de> for QueryInputValueWalker<'de> {
    type Error = InputValueSerdeError;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        match self.item {
            QueryInputValue::Null => visitor.visit_none(),
            QueryInputValue::String(s) => visitor.visit_borrowed_str(s),
            QueryInputValue::EnumValue(id) => visitor.visit_borrowed_str(self.schema_walker.walk(*id).name()),
            QueryInputValue::Int(n) => visitor.visit_i32(*n),
            QueryInputValue::BigInt(n) => visitor.visit_i64(*n),
            QueryInputValue::U64(n) => visitor.visit_u64(*n),
            QueryInputValue::Float(n) => visitor.visit_f64(*n),
            QueryInputValue::Boolean(b) => visitor.visit_bool(*b),
            QueryInputValue::List(ids) => {
                SeqDeserializer::new(self.operation[*ids].iter().map(|value| self.walk(value))).deserialize_any(visitor)
            }
            QueryInputValue::InputObject(ids) => {
                MapDeserializer::new(
                    self.operation[*ids]
                        .iter()
                        .filter_map(|(input_value_definition_id, value)| {
                            let value = self.walk(value);
                            // https://spec.graphql.org/October2021/#sec-Input-Objects.Input-Coercion
                            if value.is_undefined() {
                                None
                            } else {
                                Some((self.schema_walker.walk(*input_value_definition_id).name(), value))
                            }
                        }),
                )
                .deserialize_any(visitor)
            }
            QueryInputValue::Map(ids) => MapDeserializer::new(
                self.operation[*ids]
                    .iter()
                    .map(|(key, value)| (key.as_str(), self.walk(value))),
            )
            .deserialize_any(visitor),
            QueryInputValue::DefaultValue(id) => self
                .schema_walker
                .walk(&self.schema_walker.as_ref()[*id])
                .deserialize_any(visitor),
            QueryInputValue::Variable(id) => self.walk(*id).deserialize_any(visitor),
        }
    }

    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        if matches!(self.item, QueryInputValue::Null) {
            visitor.visit_none()
        } else {
            visitor.visit_some(self)
        }
    }

    fn deserialize_ignored_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where
        V: Visitor<'de>,
    {
        visitor.visit_unit()
    }

    forward_to_deserialize_any! {
        bool i8 i16 i32 i64 i128 u8 u16 u32 u64 u128 f32 f64 char str string
        bytes byte_buf unit unit_struct newtype_struct seq tuple
        tuple_struct map struct enum identifier
    }
}

impl<'de> IntoDeserializer<'de, InputValueSerdeError> for QueryInputValueWalker<'de> {
    type Deserializer = Self;
    fn into_deserializer(self) -> Self::Deserializer {
        self
    }
}
