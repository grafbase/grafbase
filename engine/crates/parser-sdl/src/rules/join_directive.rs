use std::collections::{BTreeSet, HashMap, HashSet};

use engine_parser::{
    parse_field,
    types::{ConstDirective, Field, InputValueDefinition},
    Positioned,
};
use engine_value::{Name, Value};
use registry_v2::{resolvers::join::JoinResolver, FieldSet};
use serde::de::Error;

use super::{
    directive::Directive,
    visitor::{VisitorContext, Warning},
};
use crate::{directive_de::parse_directive, schema_coord::SchemaCoord};

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct JoinDirective {
    pub select: FieldSelection,
}

#[derive(Debug, Clone)]
pub struct FieldSelection {
    selections: Vec<Selection>,
    variables_used: Vec<String>,
}

#[derive(Clone, Debug)]
struct Selection {
    field_name: String,
    arguments: Vec<(Name, Value)>,
}

impl Directive for JoinDirective {
    fn definition() -> String {
        "
        directive @join(select: FieldSelection!) on FIELD_DEFINITION
        "
        .into()
    }
}

impl JoinDirective {
    pub fn from_directives(
        directives: &[Positioned<ConstDirective>],
        ctx: &mut VisitorContext<'_>,
    ) -> Option<JoinDirective> {
        let directive = directives.iter().find(|directive| directive.name.node == "join")?;

        match parse_directive::<Self>(directive, &HashMap::new()) {
            Ok(directive) => Some(directive),
            Err(error) => {
                ctx.append_errors(vec![error]);
                None
            }
        }
    }

    pub fn validate_arguments(
        &self,
        arguments: &[Positioned<InputValueDefinition>],
        coord: SchemaCoord<'_>,
    ) -> Vec<Warning> {
        let argument_names = arguments
            .iter()
            .map(|argument| argument.node.name.node.as_str())
            .collect::<BTreeSet<_>>();

        let variable_names = self
            .select
            .variables_used
            .iter()
            .map(|name| name.as_str())
            .collect::<BTreeSet<_>>();

        argument_names
            .difference(&variable_names)
            .map(|unused_argument| Warning::ArgumentNotUsedByJoin(unused_argument.to_string(), coord.into_owned()))
            .collect()
    }
}

impl FieldSelection {
    pub fn required_fieldset(&self, arguments: &[Positioned<InputValueDefinition>]) -> Option<FieldSet> {
        if self.variables_used.is_empty() {
            return None;
        }

        let arguments = arguments
            .iter()
            .map(|argument| argument.node.name.node.as_str())
            .collect::<HashSet<_>>();

        let mut selections = self
            .variables_used
            .iter()
            .filter(|field| !arguments.contains(field.as_str()))
            .map(|field| registry_v2::Selection {
                field: field.clone(),
                selections: vec![],
            })
            .peekable();

        // If there are no selections then all of the variables referred to arguments
        // and we don't need to require a fieldset
        selections.peek()?;

        Some(FieldSet::new(selections))
    }

    pub fn to_join_resolver(&self) -> JoinResolver {
        JoinResolver::new(
            self.selections
                .iter()
                .map(|selection| (selection.field_name.clone(), selection.arguments.clone())),
        )
    }
}

impl<'de> serde::Deserialize<'de> for FieldSelection {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let select = String::deserialize(deserializer)?;

        let field = parse_field(select).map_err(|error| D::Error::custom(format!("Could not parse join: {error}")))?;

        validate_field(&field).map_err(D::Error::custom)?;

        let mut required_fields = BTreeSet::new();
        let mut next_field = Some(&field.node);
        while let Some(current_field) = next_field {
            required_fields.extend(
                current_field
                    .arguments
                    .iter()
                    .flat_map(|(_, value)| value.variables_used())
                    .map(|variable| variable.to_string()),
            );
            next_field = current_field.selection().next().and_then(|selection| match &selection {
                engine_parser::types::Selection::Field(field) => Some(&field.node),
                _ => None,
            });
        }

        let required_fields = required_fields.into_iter().collect::<Vec<_>>();

        Ok(FieldSelection {
            selections: SelectionIter {
                field: Some(field.node),
            }
            .collect(),
            variables_used: required_fields,
        })
    }
}

fn validate_field(mut field: &Positioned<Field>) -> Result<(), String> {
    loop {
        match field.node.selection_set.node.items.as_slice() {
            [] => return Ok(()),
            [item] => match &item.node {
                engine_parser::types::Selection::Field(inner_field) => {
                    field = inner_field;
                    continue;
                }
                _ => return Err("joins can't make use of spreads".into()),
            },
            _ => return Err("joins can contain at most a single field per selection set".into()),
        }
    }
}

struct SelectionIter {
    field: Option<Field>,
}

impl Iterator for SelectionIter {
    type Item = Selection;

    fn next(&mut self) -> Option<Self::Item> {
        let field = std::mem::take(&mut self.field)?;

        let next = Selection {
            field_name: field.name.to_string(),
            arguments: field
                .arguments
                .into_iter()
                .map(|(name, value)| (name.node, value.node))
                .collect(),
        };

        self.field = field
            .selection_set
            .node
            .items
            .into_iter()
            .next()
            .and_then(|selection| match selection.node {
                engine_parser::types::Selection::Field(inner_field) => Some(inner_field.node),
                _ => None,
            });

        Some(next)
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, assert_json_snapshot};
    use serde::Deserialize;
    use serde_json::json;

    use super::*;
    use crate::tests::assert_validation_error;

    #[test]
    fn join_happy_path() {
        let schema = r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String): String! @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "blah(id: $id)")
            }
        "#;

        let registry = crate::to_parse_result_with_variables(schema, &HashMap::new())
            .unwrap()
            .registry;

        let resolver = &registry.types["User"].fields().as_ref().unwrap()["nickname"].resolver;

        assert_json_snapshot!(resolver, @r###"
        {
          "J": {
            "fields": [
              [
                "blah",
                [
                  [
                    "id",
                    {
                      "$var": "id"
                    }
                  ]
                ]
              ]
            ]
          }
        }
        "###);
    }

    #[test]
    fn nested_join_with_arguments_and_such() {
        let schema = r#"
            extend type Query {
                greetings(name: String!): Greetings @resolver(name: "greetings")
                user: User! @resolver(name: "user")
            }

            type User {
                id: ID!
                name: String!
                greeting: String! @join(
                    select: "greetings(name: $name) { forTimeOfDay(id: $id, timeOfDay: \"morning\") }"
                )
            }

            type Greetings {
                forTimeOfDay(id: String!, timeOfDay: String!): String! @resolver(name: "timeOfDayGreeting")
            }
        "#;

        let registry = crate::to_parse_result_with_variables(schema, &HashMap::new())
            .unwrap()
            .registry;

        let resolver = &registry.types["User"].fields().as_ref().unwrap()["greeting"].resolver;

        assert_json_snapshot!(resolver, @r###"
        {
          "J": {
            "fields": [
              [
                "greetings",
                [
                  [
                    "name",
                    {
                      "$var": "name"
                    }
                  ]
                ]
              ],
              [
                "forTimeOfDay",
                [
                  [
                    "id",
                    {
                      "$var": "id"
                    }
                  ],
                  [
                    "timeOfDay",
                    "morning"
                  ]
                ]
              ]
            ]
          }
        }
        "###);

        let requires = &registry.types["User"].fields().as_ref().unwrap()["greeting"].requires;

        assert_json_snapshot!(requires, @r###"
        [
          {
            "field": "id",
            "selections": []
          },
          {
            "field": "name",
            "selections": []
          }
        ]
        "###);
    }

    #[test]
    fn join_with_missing_required_argument() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String!): String! @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "blah(id: $id)")
            }
            "#,
            "User.nickname is trying to join with Query.blah, but does not provide the non-nullable argument name"
        );
    }

    #[test]
    fn join_with_list_in_path() {
        assert_validation_error!(
            r#"
            extend type Query {
                greetings(name: String!): [Greetings] @resolver(name: "greetings")
                user: User! @resolver(name: "user")
            }

            type User {
                id: ID!
                name: String!
                greeting: String! @join(
                    select: "greetings(name: $name) { forTimeOfDay(id: $id, timeOfDay: \"morning\") }"
                )
            }

            type Greetings {
                forTimeOfDay(id: String!, timeOfDay: String!): String! @resolver(name: "timeOfDayGreeting")
            }
            "#,
            "The join on User.greeting passes through Query.greetings, which is a list.  This is not supported"
        );
    }

    #[test]
    fn join_with_nested_multiple_fields() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String!): String! @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "blah(id: $id) { foo bar }")
            }
            "#,
            "joins can contain at most a single field per selection set"
        );
    }

    #[test]
    fn join_on_missing_field() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "blah(id: $id)")
            }
            "#,
            "User.nickname is trying to join with a field named blah, which doesn't exist on the Query type"
        );
    }

    #[test]
    fn join_through_scalar() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                foo: String! @resolver(name: "foo")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "foo { bar }")
            }
            "#,
            "The join on User.nickname tries to select children of String, but String is not a composite type"
        );
    }

    #[test]
    fn join_with_a_spread() {
        assert_validation_error!(
            r#"
            extend type Query {
                greetings(name: String!): Greetings @resolver(name: "greetings")
                user: User! @resolver(name: "user")
            }

            type User {
                id: ID!
                name: String!
                greeting: String! @join(
                    select: """
                        greetings(name: $name) {
                            ... on Greetings {
                                forTimeOfDay(id: $id, timeOfDay: "morning")
                            }
                        }
                    """
                )
            }

            type Greetings {
                forTimeOfDay(id: String!, timeOfDay: String!): String! @resolver(name: "timeOfDayGreeting")
            }
            "#,
            "joins can't make use of spreads"
        );
    }

    #[test]
    fn join_with_return_type_mismatch() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!): Int @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "blah(id: $id)")
            }
            "#,
            "User.nickname is trying to join with Query.blah, but those fields do not have compatible types: 'Int' and 'String!'"
        );
    }

    #[test]
    fn join_with_variable_that_doesnt_exist() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!): String! @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "blah(id: $whatever)")
            }
            "#,
            "User.nickname declares that it requires the field whatever on User but that field doesn't exist"
        );
    }

    #[test]
    fn acceptable_return_type_nullability_mismatch() {
        let schema = r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String): String! @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String @join(select: "blah(id: $id)")
            }
            "#;

        let registry = crate::to_parse_result_with_variables(schema, &HashMap::new())
            .unwrap()
            .registry;

        let resolver = &registry.types["User"].fields().as_ref().unwrap()["nickname"].resolver;

        assert_json_snapshot!(resolver);
    }

    #[test]
    fn return_type_nullability_mismatch() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String): String @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String! @join(select: "blah(id: $id)")
            }
            "#,
            "User.nickname is trying to join with Query.blah, but those fields do not have compatible types: 'String' and 'String!'"
        );
    }

    #[test]
    fn return_type_lists_success() {
        let schema = r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String): [String!]! @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: [String] @join(select: "blah(id: $id)")
            }
            "#;

        let registry = crate::to_parse_result_with_variables(schema, &HashMap::new())
            .unwrap()
            .registry;

        let resolver = &registry.types["User"].fields().as_ref().unwrap()["nickname"].resolver;

        assert_json_snapshot!(resolver);
    }

    #[test]
    fn return_list_nullability_mismatch() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String): [String] @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: [String]! @join(select: "blah(id: $id)")
            }
            "#,
            "User.nickname is trying to join with Query.blah, but those fields do not have compatible types: '[String]' and '[String]!'"
        );
    }

    #[test]
    fn return_list_mismatch() {
        assert_validation_error!(
            r#"
            extend schema @federation(version: "2.3")

            extend type Query {
                blah(id: String!, name: String): [String] @resolver(name: "blah")
            }

            type User @key(fields: "id", resolvable: false) {
                id: ID!
                nickname: String @join(select: "blah(id: $id)")
            }
            "#,
            "User.nickname is trying to join with Query.blah, but those fields do not have compatible types: '[String]' and 'String'"
        );
    }

    #[test]
    fn join_directive_deser() {
        let directive = JoinDirective::deserialize(json!({"select": "findUser(name: $name, filters: {eq: $filters})"}));

        assert_debug_snapshot!(directive, @r###"
        Ok(
            JoinDirective {
                select: FieldSelection {
                    selections: [
                        Selection {
                            field_name: "findUser",
                            arguments: [
                                (
                                    Name(
                                        "name",
                                    ),
                                    Variable(
                                        Name(
                                            "name",
                                        ),
                                    ),
                                ),
                                (
                                    Name(
                                        "filters",
                                    ),
                                    Object(
                                        {
                                            Name(
                                                "eq",
                                            ): Variable(
                                                Name(
                                                    "filters",
                                                ),
                                            ),
                                        },
                                    ),
                                ),
                            ],
                        },
                    ],
                    variables_used: [
                        "filters",
                        "name",
                    ],
                },
            },
        )
        "###);
    }
}
