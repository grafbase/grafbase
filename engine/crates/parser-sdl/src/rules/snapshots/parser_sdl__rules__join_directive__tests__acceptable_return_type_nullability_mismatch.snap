---
source: engine/crates/parser-sdl/src/rules/join_directive.rs
expression: resolver
---
{
  "J": {
    "fields": [
      [
        "blah",
        [
          [
            "id",
            {
              "$var": "id"
            }
          ]
        ]
      ]
    ]
  }
}
